[ICE]: `TraitPredicate(...)` has escaping bound vars, so it cannot be wrapped in a dummy binder

### Code

The code is located at https://github.com/Thomasdezeeuw/heph/tree/main/rt, however I was making changes and I don't have the exact code that triggered this ICE.


### Meta

`rustc --version --verbose`:
```
rustc 1.99.0-nightly (f10db292a 2026-07-07)
binary: rustc
commit-hash: f10db292a3733b5c67c8da8c7661195ff4b05774
commit-date: 2026-07-07
host: x86_64-unknown-linux-gnu
release: 1.99.0-nightly
LLVM version: 22.1.8
```

### Error output


<details><summary><strong>Error ouput</strong></summary>
<p>

```
    Checking heph-rt v0.5.0 (/home/thomas/src/heph/rt)
error[E0277]: the trait bound `Arc<shared::RuntimeInternals>: SharedRuntimeData` is not satisfied
   --> rt/src/lib.rs:627:26
    |
627 |         spawn::try_spawn(self.internals.shared(), supervisor, new_actor, arg, options)
    |         ---------------- ^^^^^^^^^^^^^^^^^^^^^^^ the trait `SharedRuntimeData` is not implemented for `Arc<shared::RuntimeInternals>`
    |         |
    |         required by a bound introduced by this call
    |
help: the trait `SharedRuntimeData` is implemented for `shared::RuntimeInternals`
   --> rt/src/shared.rs:234:1
    |
234 | impl SharedRuntimeData for RuntimeInternals {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: required by a bound in `try_spawn`
   --> rt/src/spawn/mod.rs:168:9
    |
160 | pub(crate) fn try_spawn<RT, S, NA>(
    |               --------- required by a bound in this function
...
168 |     RT: SharedRuntimeData,
    |         ^^^^^^^^^^^^^^^^^ required by this bound in `try_spawn`


thread 'rustc' (13058) panicked at /rustc-dev/f10db292a3733b5c67c8da8c7661195ff4b05774/compiler/rustc_middle/src/ty/predicate.rs:440:9:
`TraitPredicate(<access::ThreadSafe as std::convert::From<&'a shared::RuntimeInternals>>, polarity:Positive)` has escaping bound vars, so it cannot be wrapped in a dummy binder.
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: <rustc_middle::ty::predicate::Predicate as rustc_type_ir::upcast::UpcastFrom<rustc_middle::ty::context::TyCtxt, rustc_type_ir::predicate::TraitPredicate<rustc_middle::ty::context::TyCtxt>>>::upcast_from
   3: <rustc_infer::traits::Obligation<rustc_middle::ty::predicate::Predicate>>::with_depth::<rustc_type_ir::predicate::TraitPredicate<rustc_middle::ty::context::TyCtxt>>
   4: <rustc_trait_selection::error_reporting::TypeErrCtxt>::has_impl_for_type
   5: <rustc_trait_selection::error_reporting::TypeErrCtxt>::find_explicit_cast_type
   6: <rustc_trait_selection::error_reporting::TypeErrCtxt>::report_selection_error
   7: <rustc_trait_selection::error_reporting::TypeErrCtxt>::report_fulfillment_errors
   8: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_call
   9: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  10: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_block
  11: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  12: rustc_hir_typeck::check::check_fn
  13: rustc_hir_typeck::typeck_with_inspect::{closure#0}
      [... omitted 1 frame ...]
  14: <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis::check_crate::{closure#1}>::{closure#0}
  15: rustc_hir_analysis::check_crate
  16: rustc_interface::passes::analysis
  17: rustc_query_impl::execution::try_execute_query::<rustc_middle::query::caches::SingleCache<rustc_middle::query::erase::ErasedData<[u8; 0]>>, true>
  18: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: the compiler unexpectedly panicked. This is a bug

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/home/thomas/src/heph/rustc-ice-2026-07-08T17_27_55-13056.txt` to your bug report

note: rustc 1.99.0-nightly (f10db292a 2026-07-07) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C embed-bitcode=no -C debuginfo=2 -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [typeck_root] type-checking `<impl at rt/src/lib.rs:609:1: 614:23>::try_spawn`
#1 [analysis] running analysis passes on crate `heph_rt`
end of query stack
For more information about this error, try `rustc --explain E0277`.
```

</p>
</details>

--- a/compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs
@@ -285,27 +285,21 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                             || self.tcx.is_diagnostic_item(sym::TryFrom, trait_def_id))
                             && (self.tcx.is_diagnostic_item(sym::From, leaf_trait_def_id)
                                 || self.tcx.is_diagnostic_item(sym::TryFrom, leaf_trait_def_id))
-                        {
-                            let trait_ref = leaf_trait_predicate.skip_binder().trait_ref;
-
-                            if let Some(found_ty) =
+                            && let Some(trait_ref) =
+                                leaf_trait_predicate.no_bound_vars().map(|pred| pred.trait_ref)
+                            && let Some(found_ty) =
                                 trait_ref.args.get(1).and_then(|arg| arg.as_type())
-                            {
-                                let ty = main_trait_predicate.skip_binder().self_ty();
+                            && let Some(ty) =
+                                main_trait_predicate.no_bound_vars().map(|pred| pred.self_ty())
+                            && let Some(cast_ty) =
+                                self.find_explicit_cast_type(obligation.param_env, found_ty, ty)
+                        {
+                            let found_ty_str = self.tcx.short_string(found_ty, &mut long_ty_file);
+                            let cast_ty_str = self.tcx.short_string(cast_ty, &mut long_ty_file);
 
-                                if let Some(cast_ty) =
-                                    self.find_explicit_cast_type(obligation.param_env, found_ty, ty)
-                                {
-                                    let found_ty_str =
-                                        self.tcx.short_string(found_ty, &mut long_ty_file);
-                                    let cast_ty_str =
-                                        self.tcx.short_string(cast_ty, &mut long_ty_file);
-
-                                    err.help(format!(
-                                        "consider casting the `{found_ty_str}` value to `{cast_ty_str}`",
-                                    ));
-                                }
-                            }
+                            err.help(format!(
+                                "consider casting the `{found_ty_str}` value to `{cast_ty_str}`",
+                            ));
                         }
 
                         *err.long_ty_path() = long_ty_file;

--- a/tests/ui/traits/explicit-reference-cast-hrtb.rs
+++ b/tests/ui/traits/explicit-reference-cast-hrtb.rs
@@ -1,0 +1,14 @@
+//! Regression test for #158967
+
+struct Foo;
+
+fn f()
+where
+    for<'a> Foo: From<&'a String>,
+{
+}
+
+fn main() {
+    f();
+    //~^ ERROR the trait bound `for<'a> Foo: From<&'a String>` is not satisfied [E0277]
+}

--- a/tests/ui/traits/explicit-reference-cast-hrtb.stderr
+++ b/tests/ui/traits/explicit-reference-cast-hrtb.stderr
@@ -1,0 +1,23 @@
+error[E0277]: the trait bound `for<'a> Foo: From<&'a String>` is not satisfied
+  --> $DIR/explicit-reference-cast-hrtb.rs:12:5
+   |
+LL |     f();
+   |     ^^^ unsatisfied trait bound
+   |
+help: the trait `for<'a> From<&'a String>` is not implemented for `Foo`
+  --> $DIR/explicit-reference-cast-hrtb.rs:3:1
+   |
+LL | struct Foo;
+   | ^^^^^^^^^^
+note: required by a bound in `f`
+  --> $DIR/explicit-reference-cast-hrtb.rs:7:18
+   |
+LL | fn f()
+   |    - required by a bound in this function
+LL | where
+LL |     for<'a> Foo: From<&'a String>,
+   |                  ^^^^^^^^^^^^^^^^ required by this bound in `f`
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0277`.

